### ISSUE TITLE
Move `FromRequest` and `IntoResponse` into their own crate?

### ISSUE THREAD

**Issue by davidpdrsn** [tokio-rs/axum#561]
Most crates that provide integrations with axum only need the `FromRequest` and `IntoResponse` traits. Examples include [maud](https://crates.io/crates/maud), [tower-request-id](https://crates.io/crates/tower-request-id), [tower-cookies](https://crates.io/crates/tower-cookies), [axum-flash](https://crates.io/crates/axum-flash), [and others](https://crates.io/crates/axum/reverse_dependencies).

These traits are somewhat more stable than the rest of axum. They haven't changed since 0.2 back in August, and also wont change in 0.4. So it is inconvenient for libraries since they have to bump their axum version to pull in changes that are essentially irrelevant, such as routing.

Because of this I think we should consider moving `FromRequest` and `IntoResponse` to their own crate alongside implementations for non-local types such as the strings from `http` and `bytes`. The traits would be re-exported from axum in the location they exist today.

I think we should be quite conservative about what goes into this new crate. For example I wouldn't move `Html` and `Json` even though those might be useful to some libraries. Ideally only what _has_ to be moved should be moved.

Basically the same pattern as `futures-core` and `tower-service` and many other crates I'm sure.

I suggest naming this crate `axum-core`.

@jplatte thoughts?

**Comment by kellytk** [comment]
I've two packages that extend axum.

* The first uses `async_trait`, `FromRequest`, and `RequestParts`.
* The second uses `IntoResponse`.

For any project that may be extended, I move user-facing types into a separate `project-types` package.

**Comment by davidpdrsn** [comment]
> I wonder whether there is some clever way to avoid that while still being compatible with hyper without (much) additional boilerplate when depending on the "full" axum crate.

Uh yeah thats worth exploring when we do this. Not depending hyper would mean loosing [`impl FromRequest<Body> for Body`](https://docs.rs/axum/0.3.4/src/axum/extract/request_parts.rs.html#295-301). I'm not sure how much its used in the wild but feels a bit unfortunate to loose.

**Comment by jplatte** [comment]
I think it wouldn't hurt too much to have `axum::extractors::Body` as `struct Body(pub hyper::body::Body);` since full `Body` extraction is uncommon enough. I'm more worried about the type parameter defaults.

**Comment by davidpdrsn** [comment]
I agree with that 👍

**Comment by davidpdrsn** [comment]
I gave it a shot in https://github.com/tokio-rs/axum/pull/564

**Comment by KaiJewson** [comment]
> It's a little bit unfortunate that the generic type default of `hyper::body::Body` will tie the version of axum-core to hyper's version

This can be resolved by tying it to a `hyper014` feature flag if you want to keep that impl.
